src/Workleap.Extensions.Mongo.Abstractions/MongoReflectionCache.cs

```diff
@@ -5,11 +5,17 @@ namespace Workleap.Extensions.Mongo;
 
 internal static class MongoReflectionCache
 {
-    private static readonly ConcurrentDictionary<Type, string> CollectionNames = new ConcurrentDictionary<Type, string>();
+    private static readonly ConcurrentDictionary<Type, string> CollectionNames = new();
+    private static readonly ConcurrentDictionary<Type, Type?> IndexProviderTypes = new();
 
     public static string GetCollectionName(Type documentType)
     {
-        if (!IsConcreteMongoDocumentType(documentType))
+        if (CollectionNames.TryGetValue(documentType, out var collectionName))
+        {
+            return collectionName;
+        }
+        
+        if (!documentType.IsConcreteMongoDocumentType())
```

Comment:
Yes and no. All entries to this class are from some form of Reflection. When we manually set something, its because it was discovered using Reflection.

Reply:
Renamed and extracted indexes